File: Nodal_Discontinuous_Galerkin_Methods_Hesthaven/nudg++/trunk/Src/Examples2D/CurvedEuler2D/CouetteBC2D.cpp
// CouetteBC2D.m
// function [Q] = CouetteBC2D(xin, yin, nxin, nyin, mapI, mapO, mapW, mapC, Q, time);
// 2007/06/30
//---------------------------------------------------------
#include "NDGLib_headers.h"
#include "CurvedEuler2D.h"


//---------------------------------------------------------
void CurvedEuler2D::CouetteBC2D
(
  const DVec& xi, 
  const DVec& yi, 
  const DVec& nxi, 
  const DVec& nyi, 
  const IVec& tmapI, 
  const IVec& tmapO, 
  const IVec& tmapW, 
  const IVec& tmapC, 
        double ti, 
        DMat& Qio
)
//---------------------------------------------------------
{
  // gmapB = concat(mapI,mapO,mapW, ...);
  // Check for no boundary faces
  if (gmapB.size() < 1) { return; }

  // function [Q] = CouetteBC2D(xin, yin, nxin, nyin, mapI, mapO, mapW, mapC, Q, time);
  // Purpose: evaluate solution for Couette flow

  // Couette flow (mach .2 at inner cylinder)
  // gamma = 1.4;

  int Nr = Qio.num_rows(); DVec rho,rhou,rhov,Ener;

  // wrap current state data (columns of Qio)
  rho.borrow (Nr, Qio.pCol(1));
  rhou.borrow(Nr, Qio.pCol(2));
  rhov.borrow(Nr, Qio.pCol(3));
  Ener.borrow(Nr, Qio.pCol(4));

  // update boundary nodes of Qio with boundary data
  // pre-calculated in function MapGaussFaceData()

  rho (gmapB) = rhoB;
  rhou(gmapB) = rhouB;
  rhov(gmapB) = rhovB;
  Ener(gmapB) = EnerB;
}

File: Nodal_Discontinuous_Galerkin_Methods_Hesthaven/Codes1.1/CFD2D/CouetteBC2D.m
function [Q] = CouetteBC2D(xin, yin, nxin, nyin, mapI, mapO, mapW, mapC, Q, time);

% function [Q] = CouetteBC2D(xin, yin, nxin, nyin, mapI, mapO, mapW, mapC, Q, time);
% Purpose: evaluate solution for Couette flow

% Couette flow (mach .2 at inner cylinder)
gamma = 1.4;

% extract conserved variables
rho = Q(:,:,1); rhou = Q(:,:,2); rhov = Q(:,:,3); Ener = Q(:,:,4);

mapB = [mapI;mapO];

rad = sqrt(xin(mapB).^2 + yin(mapB).^2);
theta = atan2(yin(mapB), xin(mapB));

utheta = (-rad + 16./rad)/75;
p = 1 + (1./(75^2))*( (rad.^2)/2 - 32*log(rad) - 128./rad.^2 );

rho (mapB) = 1;
rhou(mapB) = -sin(theta).*utheta;
rhov(mapB) =  cos(theta).*utheta;
Ener(mapB) = p/(gamma-1) + 0.5*(rhou(mapB).^2 + rhov(mapB).^2)./rho(mapB);

% pack modified conserved variables
Q(:,:,1) = rho; Q(:,:,2) = rhou; Q(:,:,3) = rhov; Q(:,:,4) = Ener;
return
